src/wasm-interpreter.h

```diff
@@ -4022,11 +4022,15 @@ class ModuleRunnerBase : public ExpressionRunner<SubType> {
 
     const auto& seg = *wasm.getDataSegment(curr->segment);
     auto elemBytes = element.getByteSize();
-    auto end = offset + size * elemBytes;
-    if ((size != 0ull && droppedDataSegments.count(curr->segment)) ||
-        end > seg.data.size()) {
+    uint64_t end;
+    if (std::ckd_add(&end, offset, size * elemBytes) || end > seg.data.size()) {
       trap("out of bounds segment access in array.new_data");
     }
+    uint64_t sum;
+    if (std::ckd_add(&sum, offset, size) ||
+        (sum > 0 && droppedDataSegments.count(curr->segment))) {
```

Comment:
No need to check for overflow twice
```suggestion
    if (droppedDataSegments.count(curr->segment) && end > 0) {
```